### ISSUE TITLE
Workshop on low-level interfaces with libraries

### ISSUE THREAD

**Issue by nthiery** [OpenDreamKit/OpenDreamKit#251]
This issue is to discuss a tentative workshop next Spring (April), possibly in Cernay near Paris.

TODO: discuss and refine this would-be description

(Math) Computational systems face a tension between using high level languages (e.g. Python) for expressivity, ease of use and prototyping,  and low-level languages (e.g. C/C++) for power and speed, and also for modularity (using existing libraries, or writing reusable ones). To resolve this tension, many approaches have been explored in the recent years, and the frontier between the two worlds is becoming increasingly blury.

The purpose of this workshop is to bring together a variety of developers to share expertise, seek collaboration venues, and get into concrete action:
- Developers of "binding systems" such as - [cppyy](http://doc.pypy.org/en/latest/cppyy.html#cpython, [pybind11](https://pybind11.readthedocs.io/en/stable/)
- Developers of "compilers" such as [Cython](http://www.cython.org) or [Pythran](https://github.com/serge-sans-paille/pythran)
- Developers of interactive C++ environments (cling, xeus-cling, xtensors, ...)
- Developers of (math) libraries in a low level language that wish to make their code accessible to a variety of users (e.g. [libsemigroups](https://github.com/james-d-mitchell/libsemigroups/), or low-level combinatorics code for SageMath) 
- Developpers of systems like [SageMath](http://sagemath.org/) or [Root](https://root.cern.ch/) that make an eavy use of such technologies
- People from Julia?
- ...

**Comment by nthiery** [comment]
@videlec, @serge-sans-paille (and anyone else!) would you be up for coorganizing this workshop with me?

**Comment by SylvainCorlay** [comment]
As you know, I am very interested in this idea!

**Comment by james-d-mitchell** [comment]
Me too!

**Comment by hivert** [comment]
Hi Nicolas,

On Thu, Nov 23, 2017 at 11:29:16AM +0000, Nicolas M. Thiéry wrote:
> This issue is to discuss a tentative workshop next Spring (April), possibly
> in Cernay near Paris.

As I told you, if in Cernay, I'm willing to take care of the local
organisation.

And of course I'm very interrested in the low level interfacing with EG James
semigroup or my growing https://github.com/hivert/HPCombi project.

Florent

**Comment by sebasguts** [comment]
I would be very interested in this.

I have written several interfaces for libraries, for example PyNormaliz (Python interface to Normaliz), several interfaces from GAP to different libraries (Normaliz, polymake, recently Nauty), and I am currently working on a very tight integration between GAP and Julia (making it actually possible to make calls into both directions).

I unfortunately cannot say anything about April right now, but the earliest would be the beginning of December.

**Comment by serge-sans-paille** [comment]
@nthiery I'll be happy to help!

**Comment by videlec** [comment]
@serge-sans-paille @nthiery I would also be happy to help!

**Comment by videlec** [comment]
poll for the dates at https://framadate.org/RBpcIXSHU6p8Wcu1

**Comment by nthiery** [comment]
Glad to see the enthousiam :-)
Next steps:
- extend the list of relevant projects above
- get in touch with the people / projects that are not yet following this. When you have good contacts (E.g. @SylvainCorlay with Root / cling, right?), feel free to go ahead and get in touch with them on behalf of the organizers (putting us in CC). Please edit the first comment here so that we keep track of whom has been contacted.
Thanks!

**Comment by videlec** [comment]
And as a prerequisite

- polish the description and have a default invitation e-mail

**Comment by serge-sans-paille** [comment]
cc @aguinet for the cython <> pythran collaboration

**Comment by serge-sans-paille** [comment]
Maybe developpers of CPython like @haypo would be interested too?

**Comment by sebasguts** [comment]
@videlec How long are you planning to leave the poll open?

**Comment by serge-sans-paille** [comment]
@videlec / @nthiery for those people not working for OpenDreamKit that need to take vacation to come: do we plan a 2-days workshop ? May be 3-days?

**Comment by nthiery** [comment]
>    [1]@videlec / [2]@nthiery for those people not working for OpenDreamKit
>    that need to take vacation to come: do we plan a 2-days workshop ? May
>    be 3-days?

We typically do 5 days. But if people can't make it for the full week,
that's fine. We may just want to refine the poll later on, to focus
the start or end of week on more specific themes according to whom is
available.

Cheers,
		Nicolas

**Comment by slel** [comment]
Happy to help with little things during the workshop.

**Comment by sebasguts** [comment]
Hi all,
did you already settle for some dates?
Cheers,
Sebastian

**Comment by SylvainCorlay** [comment]
I should cc @JohanMabille here who is also very interested in the subject.

**Comment by SylvainCorlay** [comment]
cc @ax3l @barche

**Comment by nthiery** [comment]
Indeed, we should fix dates very soon. Let's say before February 1rst. I am going to send now announcements and invitations to:
- sage-devel
- ...

**Comment by Axel-Naumann** [comment]
Pending details we could participate with expertise on ROOT, cling and PyROOT/cppyy. Let us know!

**Comment by nthiery** [comment]
Dear @Axel-Naumann: that would be great! Please join if you can.
(and fill in the date poll)

**Comment by SylvainCorlay** [comment]
@Axel-Naumann @vgvassilev it would be great to meet you in person at the workshop.

**Comment by SylvainCorlay** [comment]
cc @dean0x7d @maartebreddels @wjakob

**Comment by vgvassilev** [comment]
@SylvainCorlay thanks for pinging me. I signed up, I should be able to come for some of the days.

**Comment by jdemeyer** [comment]
Can we fix the dates please?

**Comment by james-d-mitchell** [comment]
It'd be great if a date could be fixed soon.

**Comment by nthiery** [comment]
Working on it ...

**Comment by SylvainCorlay** [comment]
@xhochy you might be interested in this.

**Comment by nthiery** [comment]
Thanks everyone for filling the date poll. Based on that, the workshop will happen April 30th-May 4th.
This was a tough decision to take, and I appologize in advance for those who won't be able to attend that week. We will setup soon a web page for the workshop. For now we will use the top comment on this issue. You may want to checkout the web page of the last workshop we organized at the same place for practicalities: https://wiki.sagemath.org/days85.

**Comment by jdemeyer** [comment]
Added to http://opendreamkit.org/events_activities/calendar/

**Comment by videlec** [comment]
Did you notice that the 1st of May is in the week April 30th-May 4th? This is a day without any public transport in France and a worldwide strike!

**Comment by jdemeyer** [comment]
Cool. So we have one free day!!

**Comment by nthiery** [comment]
> Did you notice that the 1st of May is in the week April 30th-May
> 4th? This is a day without any public transport in France and a
> worldwide strike!

Good point. Well hopefully we won't need public transport in the
middle of the week, and we will celebrate the workers day with ...
hard work!

For information: the premises are now reserved. We will be enjoying
"La Tour", "Le petit prince" and "Le grenier royal". That makes room
for 20+ people. And we can always host more people camping.

Next step: we should urgently advertise the final dates and keep
inviting key people.

**Comment by SylvainCorlay** [comment]
cc @davidrey @tkloczko - this is the GitHub issue on the workshop I was telling you about.

**Comment by nthiery** [comment]
I added a participants list to the page above. To register, please all follow the "Participants" link and edit the pad, taking my entry as template.

**Comment by SergeStinckwich** [comment]
Unfortunately I will not be able to join ...

**Comment by bluehood** [comment]
Hi, I believe @vgvassilev and I will participate on behalf of the ROOT team

**Comment by saraedum** [comment]
@nthiery Following your entry as a template in the participants list is a bit confusing as you are staying from 28th to 5th…but above it says that the workshop is 30th to 4th. Did I miss something obvious or should you comment on that maybe?

**Comment by nthiery** [comment]
>    [1]@nthiery Following your entry as a template in the participants list
>    is a bit confusing as you are staying from 28th to 5th…but above it
>    says that the workshop is 30th to 4th. Did I miss something obvious or
>    should you comment on that maybe?

See the item I added yesterday to the practical information: we rent
the house any way Saturday-Saturday, so people are welcome to stay
longer.

Cheers,
                                        Nicolas

**Comment by vgvassilev** [comment]
@nthiery, unfortunately I won't be able to make it. Sorry for the late cancellation.

**Comment by ax3l** [comment]
Just a note that we would love to join but my year is tightly packed with PhD activities (aka: should be writing)

**Comment by SylvainCorlay** [comment]
For those who would be looking for other accommodation options, the nearest hotel is the Abbaye de Cernay, which is a few kilometers away from the farm.

https://www.abbayedecernay.com/nos-chambres/

**Comment by SylvainCorlay** [comment]
> Just a note that we would love to join but my year is tightly packed with PhD activities (aka: should be writing).

@ax3l in the contrary it is urgent to procrastinate!

**Comment by aguinet** [comment]
Hello all!

As @serge-sans-paille , as a contributor of the cython/pythran integration, I'd be happy to join you! I'll be there on the 30/04 and 01/05. I live in Paris and will come by public transport/car/bike (depending on what's possible to do :)), thus I don't need any accomodation!

**Comment by bluehood** [comment]
Hi,
cppyy's author lets me know that the correct link to cppyy-related things is http://cppyy.readthedocs.io/en/latest/

**Comment by nthiery** [comment]
Fixed!

**Comment by dimpase** [comment]
perhaps adding (lib)semigroups to Sage ought to feature somewhere.
https://trac.sagemath.org/ticket/27295
https://trac.sagemath.org/ticket/27396
